### DISCUSSION TITLE
A quick question about the terms "module," "pre-processor," and "model"

### DISCUSSION THREAD

**Discussion by dayunbao** [Mikubill/sd-webui-controlnet#1700] [Q&A] [answered]
Hi,

First, let me thank @lllyasviel and @Mikubill for creating and maintaining such great free, open-source software!

I'm in the process on building a comic book creation app in the Godot game engine that makes API calls to the sd-webui-controlnet server for image generation.  The app will allow users to create an image generation pipeline using something like a visual scripting interface. The basic idea is that users will be able to chain together nodes that make API calls to different models to generate a single panel of a comic book's page.  I also plan to have a website where users can share the comic books they create.  I have some ideas for how to combine all of this with custom tools built in the game engine that I think will allow more fine-grained control over image generation.  I've been testing with Stability AI's API while building the base of the interface (running a g4 instance type in AWS is not cheap), but I'm ready to start building and testing with sd-webui-controlnet.  However, I wanted to make sure that I understand some of the terminology before I begin.

I'm a bit confused about the difference between a "model" and a "pre-processor" (which I believe is the same thing that's referred to as a "module" in the API docs).  I've called both the `/controlnet/model_list` and `/controlnet/module_list` endpoints for comparison, and see some overlap.  For example,`/controlnet/model_list`  returns this
- control_v11p_sd15_openpose [cab727d4]

while `/controlnet/module_list` returns this:
- "openpose"
- "openpose_hand"
- "openpose_face"
- "openpose_faceonly"
- "openpose_full"

Are the `openpose_x` modules/pre-processors in fact functionality of the `openpose` model that has been exposed individually?

**Comment by huchenlei** [comment, marked as answer]
The preprocessor(module) takes a image and convert it into a format that the model accepts. 
If you can directly provide the format the model needs, you should choose preprocessor 'none'.

Try to use the extension more with WebUI and you should be able to get the difference.

**Comment by dayunbao** [comment]
Great.  Thanks!
